coremltools/converters/mil/mil/ops/defs/iOS15/linear.py

```diff
@@ -289,7 +289,8 @@ def type_inference(self):
         x_shape = x.shape
         y_shape = y.shape
         assert len(x_shape) == len(y_shape), "inputs not of the same rank"
-        assert x_shape[-1] == y_shape[-3], "input shapes incompatible"
+        assert x_shape[-1] == y_shape[-3] or is_symbolic(x_shape[-1]) or is_symbolic(y_shape[-3]), \
```

Comment:
Not sure this is a correct fix ...
In theory, even though the input shapes are symbolic,
they should still be aligned imo ...
Feel like this could be due to other incorrect type inference upstream?
There is a fix for that (for conv layer) in this branch:
https://github.com/apple/coremltools/pull/2027
@TobyRoseman would you mind checking if the new fix can solve your problem here?

Reply:
What do you mean by "aligned"? They should be the same symbols?

Reply:
Yes, I do think they should be the same symbol

Reply:
Using tip of `main`, we still get the error. I think we should take this fix. I'll make the change suggested by @junpeiz.